src/internal/pfsdb/commit_provenance.go

```diff
@@ -0,0 +1,214 @@
+          SELECT DISTINCT commit_id 
+          FROM pfs.commits, subv 
+          WHERE int_id = subv.from_id AND commit_set_id != $2;`
+	rows, err := tx.QueryxContext(ctx, q, id, id)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	defer rows.Close()
+	cs := make([]*pfs.Commit, 0)
+	for rows.Next() {
+		var commit string
+		if err := rows.Scan(&commit); err != nil {
+			return nil, errors.EnsureStack(err)
+		}
+		cs = append(cs, ParseCommit(commit))
+	}
+	return cs, nil
+}
+
+func AddCommit(ctx context.Context, tx *pachsql.Tx, commit *pfs.Commit) error {
+	stmt := `INSERT INTO pfs.commits(commit_id, commit_set_id, branch) VALUES ($1, $2, $3)`
+	_, err := tx.ExecContext(ctx, stmt, CommitKey(commit), commit.ID, BranchKey(commit.GetBranch()))
+	return errors.EnsureStack(err)
+}
+
+func DeleteCommit(ctx context.Context, tx *pachsql.Tx, commitKey string) error {
+	id, err := getCommitTableID(ctx, tx, commitKey)
+	if err != nil {
+		return err
+	}
+	stmt := `DELETE FROM pfs.commits WHERE int_id = $1;`
+	_, err = tx.ExecContext(ctx, stmt, id)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	stmt = `DELETE FROM pfs.commit_provenance WHERE from_id = $1 OR to_id = $1;`
+	_, err = tx.ExecContext(ctx, stmt, id)
+	return errors.EnsureStack(err)
+}
+
+func getCommitTableID(ctx context.Context, tx *pachsql.Tx, commitKey string) (int, error) {
+	query := `SELECT int_id FROM pfs.commits WHERE commit_id = $1;`
+	rows, err := tx.QueryxContext(ctx, query, commitKey)
+	if err != nil {
+		return 0, errors.EnsureStack(err)
+	}
+	var id int
+	for rows.Next() {
+		if err := rows.Scan(&id); err != nil {
+			return 0, errors.EnsureStack(err)
+		}
+	}
+	return id, nil
+}
+
+func AddCommitProvenance(ctx context.Context, tx *pachsql.Tx, from, to *pfs.Commit) error {
+	query := `SELECT int_id, commit_id FROM pfs.commits WHERE commit_id = $1 OR commit_id = $2;`
+	rows, err := tx.QueryxContext(ctx, query, CommitKey(from), CommitKey(to))
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	var fromId, toId int
+	var count int
+	for rows.Next() {
+		count++
+		var tmp int
+		var commitId string
+		if err := rows.Scan(&tmp, &commitId); err != nil {
+			return errors.EnsureStack(err)
+		}
+		if commitId == CommitKey(from) {
+			fromId = tmp
+		} else {
+			toId = tmp
+		}
+	}
+	if count != 2 {
+		return errors.Errorf("expected two existing commits, got %v", count)
+	}
+	return addCommitProvenance(ctx, tx, fromId, toId)
+}
+
+func addCommitProvenance(ctx context.Context, tx *pachsql.Tx, from, to int) error {
+	stmt := `INSERT INTO pfs.commit_provenance(from_id, to_id) VALUES ($1, $2) ON CONFLICT DO NOTHING;`
+	_, err := tx.ExecContext(ctx, stmt, from, to)
+	return errors.EnsureStack(err)
+}
+
+func SetupCommitProvenanceV0(ctx context.Context, tx *pachsql.Tx) error {
+	_, err := tx.ExecContext(ctx, schema)
+	return errors.EnsureStack(err)
+}
+
+var schema = `
+	CREATE TABLE pfs.commits (
+		int_id BIGSERIAL PRIMARY KEY,
+		commit_id VARCHAR(4096) UNIQUE,
```

Comment:
Should `commit_set_id` and `branch` have `NOT NULL`, or is there a useful value of NULL here?

Reply:
If `branch` can be `NOT NULL` then [this conditional](https://github.com/pachyderm/pachyderm/pull/8306/files#diff-4b93de8aecf18588c7903d6c07e15942c5b29a5c4ce9e4aa1a1ec058b15934d6R57-R59) can be avoided.